### DISCUSSION TITLE
How do I update parameters from the kedro notebook?

### DISCUSSION THREAD

**Discussion by ghost** [kedro-org/kedro#1882] [Idea]
Hello

I wanted to update the parameters with in the parameters.yml from the jupyter notebook every-time beore I do the session.run().
I was able to update the parameters from context as mentioned [here](https://stackoverflow.com/questions/60299478/setting-parameters-in-kedro-notebook) but when I do the session.run() the parameters are getting loaded from **catalog** rather than **context**. Can anyone please advice me how do I update the parameters with in catalog before I run the pipeline?

Thank you!!

**Comment by noklam** [comment]
The StackOverflow you linked is 2.5 years old so it's a bit outdated. Assuming you are on Kedro 0.18.3

There are two ways to do it
## 1. Use the `%reload_kedro` magic, which creates the `session` object for you
```python
%load_ext kedro.python   # Not needed if you use kedro ipython / kedro jupyter as it will do it for you
%reload_kedro --env=<env> --params=<params>  # Similar to kedro run CLI 
session.run()
```

## 2. Create the session manually

```
from kedro.framework.session import KedroSession
with KedroSession.create(package_name=None, project_path=None, save_on_close=True, env=None, extra_params=None):
    session.run()
```

tl;dr You can pass the extra_params when creating the `session`.

The detailed answer is, a new `context` is created every time you do `session.run`, as `session` only allows 1 run now, you cannot reuse the same session.

**Reply by ghost** [reply]
Hi @noklam 
Yes, I'm using Kedro 0.18.3
Thank you for providing both the ways. Unfortunately both of them are not working for me
1) In the first method, if I use something like %reload_kedro --params={'n_estimators':10}, the project path is getting updated to /../notebooks/--params='n_estimators':10 and it is throwing the error

2) In the 2nd method, the parameter is still not getting updated when I do session.run(), 
so I've done in this way - 
```
with KedroSession.create(project_path='../...', extra_params={'n_estimators':10}):
    print(context.params['n_estimators'])
    session.run()
```
The print statement is still providing the initial parameter value for n_estimator rather than the updated value and also when the session runs, it is still taking the initial parameter value rather than the updated value

**Reply by noklam** [reply]
For 1., would it work if you specify the path to the root directory `../`?

Does it works if you do `kedro run --params`? My bet is not, do let me know if this is the case as it indicates inconsistency.

If it happens that you are trying to use TemplateConfigLoader but you need to override globals at runtime, check this out https://github.com/kedro-org/kedro/discussions/1782.

If none of this work, it would be great if you can document how to reproduce your issues, even better if you can just share the repository if it's public. There are not enough context to determine what's the issue.

**Reply by ghost** [reply]
Hi @noklam

if I add my root directory as ../ 
I'm still getting the earlier runtime error by not updating the project path properly (project path is still getting updated to - `"/xx/xx/xx/xx/--params='n_estimators':7"` and I get the below error
`Could not find the project configuration file 'pyproject.toml' in /xx/xx/xx/xx/ --params='n_estimators':10.`

kedro run --params is working fine and the parameters are getting updated

If I add the ConfigLoader class, jupyter notebook is not at all working; not sure if I need to add anything else apart from that in the settings.py file

**Reply by ghost** [reply]
Hey @noklam 
Can you let me know if I'm missing something?
Thank you!!

**Reply by noklam** [reply]
Would you be able to share a repository or something that I can reproduce? It's not clear to me what's the problem here.

**Reply by ghost** [reply]
In short, I'm able to update the parameters defined in the parameters.yml file from CLI but I'm unable to update them from a jupyter notebook.

For e.g- if I run the following code - 

`with KedroSession.create(project_path='/Users/praneeth/Desktop/kedro-pilot/', extra_params={'n_estimators':10}):
    print(context.params['n_estimators'])`

I'm getting the output as **16** (which is originally defined in the parameters.yml file) rather than 10 which I'm specifying with in the extra_params and hence during the session run while model building stage, the model is taking input as 16 instead of 10

**Reply by noklam** [reply]
Did you recreate the context object? I think you are pointing to an old context that's why it isn't updated.

The ipython extension preload a few variables, you can refresh it with the %reload_kedro magic. If you choose to create a session yourself, you need to get a new context object instead of using the preloaded object.

If you can just print out the parameters within the node function, I think you should see 10.

TLDR I think the parameters is updated but you are printing the old context thus not seeing the new parameters.

**Reply by ghost** [reply]
Hi @noklam 
Even after running the %reload_kedro magic, the parameter is still being seen as 16 to the pipeline instead of 10 when the session runs. 
At the moment, I was able to do it by initiating the pipeline from the python script rather than jupyter notebook with the same exact code. Going to give up the jupyter notebook approach

Also on the lines of parameter tuning, I also have an another question. Can you let me know if there is a way to specify multiple data folders in catalog.yml?
At the moment my data storage has a root folder which has one folder corresponding to each date like - 2022-09-30, 2022-10-01... etc. and internally each date folder has many parquet files. 
I was able to load the multiple parquet files corresponding to one date folder in this way using the PartitionedDataSet- 

**type: PartitionedDataSet
path: "/Users/praneeth/Desktop/data/2022-09-30"
dataset: pandas.ParquetDataSet
load_args:
        engine: pyarrow
        use_nullable_dtypes: True**

Can you suggest me if there is a way to load the specified folders instead of a single folder? (for e.g -> to load all the folders between 2022-09-26 till 2022-10-03) 
I can create a new discussion for this if needed?

**Reply by noklam** [reply]
@praneeth5222 Would you be able to share an example by modifying a starter project like the iris or spaceflights one? What you have described is very unusual.

**Reply by ghost** [reply]
Sure, will do that
Should I raise a new discussion about specifying multiple folders in catalog.yml file?

**Reply by noklam** [reply]
@praneeth5222 yes please
